YosysHQ/nextpnr
# Gowin target

As the title suggests, it's a Gowin target for Nextpnr. It's a bit rough around the edges, but I just successfully ran a full flow and tested it on a GW1NR-9.

It's mostly a direct port from the old nexpnr-generic target, with added support for all LUTs, DFF reset/enable, correct validation checks, and pin constraints.

This means there is no deduplication and loading isn't instantaneous. When I add support for larger GW2A devices, I intend to revisit this target to add those things as needed. For the GW1N devices, it's not a real problem.

The main problem for now is that routing the attosoc demo takes forever. It has 53% utilization, but at about 8k unrouted nets it basically makes no progress at all.

For now the main things to do are

- [x] remove all debugging leftovers
- [x] figure out if attosoc routing congestion can be improved
- [x] address any feedback on this PR

### How to test

I have built this branch with the following CMake variables, where the notable thing is that the python interpreter is the same one as is used for Project Apicula. If you're using system Python, I guess these options are not needed.

```json
{
    "ARCH": "gowin",
    "BUILD_GUI": false,
    "PYTHON_EXECUTABLE": "/home/pepijn/code/apicula/env/bin/python",
    "GOWIN_BBA_EXECUTABLE": "/home/pepijn/code/apicula/env/bin/gowin_bba",
}
```

I'm pushing a new Apycula to Pypi as I write this, so you should be able to `pip install apycula` soon.

From there you should be able to use the latest yosys to `synth_gowin` an example program.
A simple way is to just `make blinky.json` in https://github.com/YosysHQ/apicula/tree/master/examples

Then you should be able to run nextpnr like so
```
nxtpnr-gowin --json apicula/examples/attosoc.json \
                --write out.json \
                --device GW1NR-9 \
                --family GW1N-9 \
                --package QN881 \
                --speed C6/I5 \
                --cst apicula/examples/gw1n9.cst
```

And then finally pack the bitstream with 

```
gowin_pack -d GW1N-9 -o blinky.json out.json
```
